### PR TITLE
o/ifacestate: don't fail remove if disconnect hook fails

### PR DESCRIPTION
**Pull request by MiguelPires**
Ignore hook failures when doing auto-disconnect so they don't prevent a snap from being removed.

https://bugs.launchpad.net/snapd/+bug/1845880

### CODE REVIEW

**Review by mardy — APPROVED** [review]
LGTM!

File: overlord/ifacestate/ifacestate_test.go [hunk 1]
```diff
@@ -4250,6 +4250,86 @@ plugs:
...
+	s.manager(c)
+	s.state.Lock()
+
+	conn := &interfaces.Connection{
+		Plug: interfaces.NewConnectedPlug(consumerInfo.Plugs["plug"], nil, nil),
+		Slot: interfaces.NewConnectedSlot(producerInfo.Slots["slot"], nil, nil),
+	}
+
+	// call disconnect with the AutoDisconnect flag (used when removing a snap)
+	ts, err := ifacestate.DisconnectPriv(s.state, conn, ifacestate.NewDisconnectOptsWithAutoSet())
```
Comment (mvo5) [inline overlord/ifacestate/ifacestate_test.go:4299]: Should we add an error task here to check that the disconnect really worked even if there is an error?

Comment (MiguelPires) [reply overlord/ifacestate/ifacestate_test.go:4299]: The change doesn't ignore errors in tasks, only in hooks, so I think adding a failed task would make the test fail (which is expected). Line 4258 of the test simulates a hook failing which shouldn't fail the run-hook task (and consequently the disconnect)

File: overlord/ifacestate/ifacestate_test.go [hunk 2]
```diff
@@ -4250,6 +4250,86 @@ plugs:
...
+	var hooksCount int
+	for _, t := range ts.Tasks() {
+		if t.Kind() != "run-hook" {
+			continue
+		}
+
+		var hooksetup hookstate.HookSetup
+		err := t.Get("hook-setup", &hooksetup)
+		c.Assert(err, IsNil)
+		c.Check(hooksetup.IgnoreError, Equals, true)
```
Comment (stolowski) [inline overlord/ifacestate/ifacestate_test.go:4320]: Could you also add similar check for "undo-hook-setup"?

**Review by stolowski — APPROVED** [review]
Looks good, one remark about test check for undo-hook-setup. Thank you

File: overlord/ifacestate/ifacestate_test.go (hunk 1 above)
Comment (stolowski) [reply overlord/ifacestate/ifacestate_test.go:4299]: I think it's fine, existing hookstate tests already cover IgnoreError flag e.g. TestHookTaskHandleIgnoreErrorWorks

File: overlord/ifacestate/ifacestate_test.go (hunk 2 above)
Comment (MiguelPires) [reply overlord/ifacestate/ifacestate_test.go]: Done. Thanks

File: overlord/ifacestate/ifacestate_test.go (hunk 1 above)
Comment (mvo5) [reply overlord/ifacestate/ifacestate_test.go:4299]: Thanks for explaining this, that's fine then :)

**Comment by mvo5** [comment]
I see a spread test failure now:
```
2021-09-21T15:59:34.5695768Z 2021-09-21 15:59:34 Restoring google:ubuntu-16.04-64:tests/main/help (sep211524-253764)...
2021-09-21T15:59:36.3078749Z ##[error]2021-09-21 15:59:36 Error executing google:ubuntu-16.04-64:tests/main/disconnect-undo (sep211524-248954) : 
2021-09-21T15:59:36.3116311Z -----
2021-09-21T15:59:36.3117336Z + MATCH 'test-disconnect:network .*:network'
2021-09-21T15:59:36.3118106Z + snap connections
2021-09-21T15:59:36.3119362Z + echo 'Disconnect fails due to failing disconnect hook'
2021-09-21T15:59:36.3120233Z Disconnect fails due to failing disconnect hook
2021-09-21T15:59:36.3121177Z + MATCH 'failure of disconnect hook'
2021-09-21T15:59:36.3122631Z + snap disconnect test-disconnect:network
2021-09-21T15:59:36.3124432Z + echo 'And network plug remains connected'
2021-09-21T15:59:36.3126282Z And network plug remains connected
2021-09-21T15:59:36.3127357Z + MATCH 'test-disconnect:network .*:network'
2021-09-21T15:59:36.3127966Z + snap connections
2021-09-21T15:59:36.3128917Z + echo 'Snap removal fails because of failing disconnect hook'
2021-09-21T15:59:36.3129719Z Snap removal fails because of failing disconnect hook
2021-09-21T15:59:36.3130622Z + MATCH 'failure of disconnect hook'
2021-09-21T15:59:36.3131777Z + snap remove test-disconnect
2021-09-21T15:59:36.3132361Z grep error: pattern not found, got:
2021-09-21T15:59:36.3133200Z test-disconnect removed
2021-09-21T15:59:36.3133777Z -----
```

**Comment by stolowski** [comment]
> I see a spread test failure now:
> 
> ```
> 2021-09-21T15:59:36.3129719Z Snap removal fails because of failing disconnect hook
> 2021-09-21T15:59:36.3130622Z + MATCH 'failure of disconnect hook'
> 2021-09-21T15:59:36.3131777Z + snap remove test-disconnect
> 2021-09-21T15:59:36.3132361Z grep error: pattern not found, got:
> 2021-09-21T15:59:36.3133200Z test-disconnect removed
> 2021-09-21T15:59:36.3133777Z -----

Right, we had a test explicitly expecting remove to fail in this case.

**Comment by stolowski** [comment]
> > I see a spread test failure now:
> > ```
> > 2021-09-21T15:59:36.3129719Z Snap removal fails because of failing disconnect hook
> > 2021-09-21T15:59:36.3130622Z + MATCH 'failure of disconnect hook'
> > 2021-09-21T15:59:36.3131777Z + snap remove test-disconnect
> > 2021-09-21T15:59:36.3132361Z grep error: pattern not found, got:
> > 2021-09-21T15:59:36.3133200Z test-disconnect removed
> > 2021-09-21T15:59:36.3133777Z -----
> > ```
> 
> Right, we had a test explicitly expecting remove to fail in this case.

That test was introduced with https://github.com/snapcore/snapd/pull/8928 to demonstrate the problem with snap removal and there was no other deeper reason for this behavior, so I think this PR is fine and the spread test should be modified.

**Pull request merged**